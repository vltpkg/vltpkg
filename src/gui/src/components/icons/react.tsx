export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
