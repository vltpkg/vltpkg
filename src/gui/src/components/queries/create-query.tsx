import { useEffect, useState } from 'react'
import { Button } from '@/components/ui/button.tsx'
import { Input } from '@/components/ui/input.tsx'
import { useGraphStore } from '@/state/index.ts'
import { v4 as uuidv4 } from 'uuid'
import type { QueryLabel, DashboardData } from '@/state/types.ts'
import { LabelSelect } from '@/components/labels/label-select.tsx'
import { LabelBadge } from '@/components/labels/label-badge.tsx'
import { useToast } from '@/components/hooks/use-toast.ts'
import { Label } from '@/components/ui/label.tsx'
import {
  Popover,
  PopoverTrigger,
  PopoverContent,
} from '@/components/ui/popover.tsx'
import {
  Tooltip,
  TooltipProvider,
  TooltipTrigger,
  TooltipContent,
} from '@/components/ui/tooltip.tsx'
import { ChevronsUpDown, CircleHelp } from 'lucide-react'
import { cn } from '@/lib/utils.ts'
import { DirectorySelect } from '@/components/directory-select.tsx'

interface CreateQueryProps {
  onClose: () => void
  className?: string
  dashboard?: DashboardData
}

export const CreateQuery = ({
  className,
  onClose,
  dashboard,
}: CreateQueryProps) => {
  const [queryName, setQueryName] = useState<string>('')
  const [query, setQuery] = useState<string>('')
  const [directory, setDirectory] = useState<string>('')
  const [labels, setLabels] = useState<QueryLabel[]>([])

  const [isCreationValid, setIsCreationValid] =
    useState<boolean>(false)
  const [_labelPopoverOpen, setLabelPopoverOpen] =
    useState<boolean>(false)

  const { toast } = useToast()
  const saveQuery = useGraphStore(state => state.saveQuery)

  useEffect(() => {
    if (queryName.trim() !== '' && query.trim() !== '') {
      setIsCreationValid(true)
    } else {
      setIsCreationValid(false)
    }
  }, [queryName, query])

  const handleCancel = () => {
    onClose()
  }

  const handleCreate = () => {
    saveQuery({
      id: uuidv4(),
      name: queryName,
      context: directory,
      query,
      dateCreated: new Date().toISOString(),
      dateModified: new Date().toISOString(),
      labels,
    })
    toast({
      title: `${queryName} saved successfully`,
    })
    onClose()
  }

  return (
    <div
      className={cn(
        'rounded-sm border border-[1px] border-muted-foreground/25 bg-card transition-all',
        className,
      )}>
      <div className="flex flex-col gap-3 px-3 py-3">
        {/* query preview */}
        <div className="flex w-full items-center">
          <p className="text-base font-medium">
            {queryName.trim() === '' ? 'New Query' : queryName}
          </p>

          <div className="ml-auto flex items-center gap-2">
            {labels.map((label, idx) => (
              <LabelBadge key={idx} {...label} />
            ))}
          </div>
        </div>

        <div className="mt-4 flex w-full gap-3">
          <div className="flex flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Name
            </Label>
            <Input
              type="text"
              value={queryName}
              onChange={e => setQueryName(e.target.value)}
              placeholder="Name"
            />
          </div>

          <div className="flex grow flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Query
            </Label>
            <Input
              type="text"
              value={query}
              onChange={e => setQuery(e.target.value)}
              placeholder="Query"
            />
          </div>

          <div className="flex w-[300px] flex-col gap-2">
            <TooltipProvider>
              <Tooltip>
                <TooltipTrigger className="inline-flex cursor-default items-center">
                  <Label className="border-none text-sm font-medium">
                    Directory (optional)
                  </Label>
                  <CircleHelp
                    className="text-muted-foreground"
                    size={18}
                  />
                </TooltipTrigger>
                <TooltipContent>
                  Set directory to 'Global' to reuse across all
                  projects.
                </TooltipContent>
              </Tooltip>
            </TooltipProvider>
            <DirectorySelect
              directory={directory}
              setDirectory={setDirectory}
              dashboard={dashboard}
            />
          </div>

          <div className="flex w-[175px] flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Labels
            </Label>
            <Popover>
              <PopoverTrigger asChild>
                <Button
                  variant="outline"
                  role="combobox"
                  className="w-full justify-between">
                  Select labels
                  <ChevronsUpDown className="opacity-50" />
                </Button>
              </PopoverTrigger>
              <PopoverContent className="p-0">
                <LabelSelect
                  setIsOpen={setLabelPopoverOpen}
                  selectedItems={labels}
                  setItems={setLabels}
                />
              </PopoverContent>
            </Popover>
          </div>
        </div>

        {/* footer */}
        <div className="mb-0.5 flex items-end justify-end gap-3">
          <Button onClick={handleCancel} variant="outline">
            Cancel
          </Button>
          <Button disabled={!isCreationValid} onClick={handleCreate}>
            Create Query
          </Button>
        </div>
      </div>
    </div>
  )
}
