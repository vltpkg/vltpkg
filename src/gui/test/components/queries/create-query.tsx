import React from 'react'
import { vi, describe, it, expect, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.ts'
import { CreateQuery } from '@/components/queries/create-query.tsx'

vi.mock('@/components/ui/button.tsx', () => ({
  Button: 'gui-button',
}))

vi.mock('@/components/ui/input.tsx', () => ({
  Input: 'gui-input',
}))

vi.mock('@/components/labels/label-select.tsx', () => ({
  LabelSelect: 'gui-label-select',
}))

vi.mock('@/components/labels/label-badge.tsx', () => ({
  LabelBadge: 'gui-label-badge',
}))

vi.mock('@/components/ui/label.tsx', () => ({
  Label: 'gui-label',
}))

vi.mock('@/components/ui/popover.tsx', () => ({
  Popover: 'gui-popover',
  PopoverContent: 'gui-popover-content',
  PopoverTrigger: 'gui-popover-trigger',
}))

vi.mock('@/components/ui/tooltip.tsx', () => ({
  Tooltip: 'gui-tooltip',
  TooltipContent: 'gui-tooltip-content',
  TooltipProvider: 'gui-tooltip-provider',
  TooltipTrigger: 'gui-tooltip-trigger',
}))

vi.mock('lucide-react', () => ({
  ChevronsUpDown: 'gui-chevron-icon',
  CircleHelp: 'gui-circle-help-icon',
}))

vi.mock('@/components/directory-select.tsx', () => ({
  DirectorySelect: 'gui-directory-select',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

describe('create-query', () => {
  const setStateMock = vi.fn()
  vi.spyOn(React, 'useState').mockReturnValue(['', setStateMock])

  it('should render correctly', () => {
    const Container = () => {
      return <CreateQuery onClose={setStateMock} />
    }

    const { container } = render(<Container />)

    expect(container.innerHTML).toMatchSnapshot()
  })
})
