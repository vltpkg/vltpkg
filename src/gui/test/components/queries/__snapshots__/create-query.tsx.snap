// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create-query > should render correctly 1`] = `

<div class="rounded-sm border-[1px] border-muted-foreground/25 bg-card transition-all">
  <div class="flex flex-col gap-3 px-3 py-3">
    <div class="flex w-full items-center">
      <p class="text-base font-medium">
        New Query
      </p>
      <div class="ml-auto flex items-center gap-2">
      </div>
    </div>
    <div class="mt-4 flex w-full gap-3">
      <div class="flex flex-col gap-2">
        <gui-label classname="border-none text-sm font-medium">
          Name
        </gui-label>
        <gui-input
          type="text"
          value
          placeholder="Name"
        >
        </gui-input>
      </div>
      <div class="flex grow flex-col gap-2">
        <gui-label classname="border-none text-sm font-medium">
          Query
        </gui-label>
        <gui-input
          type="text"
          value
          placeholder="Query"
        >
        </gui-input>
      </div>
      <div class="flex w-[300px] flex-col gap-2">
        <gui-tooltip-provider>
          <gui-tooltip>
            <gui-tooltip-trigger classname="inline-flex cursor-default items-center">
              <gui-label classname="border-none text-sm font-medium">
                Directory (optional)
              </gui-label>
              <gui-circle-help-icon
                classname="text-muted-foreground"
                size="18"
              >
              </gui-circle-help-icon>
            </gui-tooltip-trigger>
            <gui-tooltip-content>
              Set directory to 'Global' to reuse across all projects.
            </gui-tooltip-content>
          </gui-tooltip>
        </gui-tooltip-provider>
        <gui-directory-select directory>
        </gui-directory-select>
      </div>
      <div class="flex w-[175px] flex-col gap-2">
        <gui-label classname="border-none text-sm font-medium">
          Labels
        </gui-label>
        <gui-popover>
          <gui-popover-trigger aschild="true">
            <gui-button
              variant="outline"
              role="combobox"
              classname="w-full justify-between"
            >
              Select labels
              <gui-chevron-icon classname="opacity-50">
              </gui-chevron-icon>
            </gui-button>
          </gui-popover-trigger>
          <gui-popover-content classname="p-0">
            <gui-label-select selecteditems>
            </gui-label-select>
          </gui-popover-content>
        </gui-popover>
      </div>
    </div>
    <div class="mb-0.5 flex items-end justify-end gap-3">
      <gui-button variant="outline">
        Cancel
      </gui-button>
      <gui-button disabled="true">
        Create Query
      </gui-button>
    </div>
  </div>
</div>

`;
